hashdist/cli/remote_cli.py

```diff
@@ -0,0 +1,304 @@
+                    download_request.progress_listener(StdoutProgressListener())
+                    storage.download(download_request)
+                    manifest = json.loads(str(remote_manifest_string.get_bytes()))
+                except:
+                    ctx.logger.warn("Failed to get remote manifest; all packages will be pushed")
+                    manifest = {}
+                ctx.logger.info("Writing local copy of remote  manifest")
+                with open(pjoin("..","source_manifest.json"),"w") as f:
+                    f.write(json.dumps(manifest))
+                ctx.logger.info("Calculating which packages to push")
+                push_manifest = {}
+                for subdir in ['files']+[pjoin('packs',pack_type) for pack_type in ['tar.bz2', 'tar.gz', 'zip']]:
+                    if not manifest.has_key(subdir):
+                        manifest[subdir] = []
+                    for source_pack in os.listdir(pjoin(cache.cache_path,subdir)):
+                        if source_pack in manifest[subdir] and not args.force:
+                            ctx.logger.info(subdir+"/"+source_pack+" already on remote")
+                        else:
+                            if not push_manifest.has_key(subdir):
+                                push_manifest[subdir]=set()
+                            push_manifest[subdir].add(source_pack)
+                ctx.logger.info("Source packs to push"+`push_manifest`)
+                for subdir,source_packs in push_manifest.iteritems():
+                    for source_pack in source_packs:
+                        manifest[subdir].append(source_pack)
+                        source_pack_path = pjoin(subdir,source_pack)
+                        sys.stdout.write("Pushing "+`source_pack_path`+"\n")
+                        fpath = CPath('/src/'+subdir)
+                        storage.create_folder(fpath)
+                        bpath=fpath.add(source_pack)
+                        upload_request = CUploadRequest(bpath,FileByteSource(source_pack_path)) 
+                        upload_request.progress_listener(StdoutProgressListener())
+                        storage.upload(upload_request)
+                        sys.stdout.write("Syncing manifest")
+                        new_manifest_string = json.dumps(manifest)
+                        new_manifest_bytes = bytes(new_manifest_string)
+                        manifest_byte_source = MemoryByteSource(new_manifest_bytes)
+                        fpath = CPath('/src')
+                        bpath=fpath.add("source_manifest.json")
+                        upload_request = CUploadRequest(bpath,manifest_byte_source).content_type('text/plain')
+                        upload_request.progress_listener(StdoutProgressListener())
+                        storage.upload(upload_request)
+                        with open(pjoin("..","source_manifest.json"),"w") as f:
+                            f.write(new_manifest_string)
```

Comment:
This is a rather bulky function that is going to be hard to refactor in the future.  If you're looking for suggestions to improve the code, breaking this function up would improve the clarity of the code and its maintainability.